cmd/zrepl/zrepl.c

```diff
@@ -550,11 +583,10 @@ uzfs_zvol_mgmt_thread(void *arg)
 	}
 
 	while (1) {
-		bzero(&hdr, sizeof (hdr));
-		count = read(sfd, (char *)&hdr, sizeof (hdr));
-		if (count <= 0) {
-			ZREPL_ERRLOG("Replica-iSCSI Tgt connection got "
-			    "disconnected with err:%d\n", errno);
+		rc = read_handshake(sfd, &hdr);
```

Comment:
We need to support/manage many more commands here. Current code assume that it will handle just HANDSHAKE cmd/msg only. We need to have Switch statement here to handle different commands received from iSCSI tgt (Handshake is one of them, now for rebuild I am implementing many more).

Reply:
agreed. Although the situation is even a bit more complicated. Reading handshake is not like reading any other command:
 1) handshake command must come first
 2) handshake command must not repeat
 3) only first 2 bytes need to be read from handshake message before anything else is, to decide if the sender's version is supported.

Hence I would process handshake command out of the loop and then have the switch statement inside the loop as you have described, but only for all other commands except handshake. But I would make this change once we add more commands - not now when there is just one.

Reply:
hm, it actually may repeat. It should not repeat on data connection but can repeat on mgmt connection in case iscsit requests information (host:port) for multiple volumes. What do you think? In that case we should perhaps decouple handshake message from zvol query. I mean as in the following picture (that is for mgmt conn):

zrepl                        iscsit
  ------connect----->
  ------HS msg----->
  <-----HS repl------
  <----zvol query---
  -----zvol info------>
  <----zvol query---
  -----zvol info----->
             ....

That would eliminate one of the current weirdnesses of the protocol when the client (zrepl) which connects to iscsi, expects iscsit to send the first msg on connection. Now it would be the zrepl who sends HS over the wire and iscsit (server from this point of view) replies to the handshake.

Reply:
That can be done i.e. first message from Zrepl to iSCSI controller. But here HS is meant for finding volume which iSCSI know off. Zreplica search that volume and reply back to controller with appropriate response. If Replica has to send first message, I am not what info it should send. I think we should handle HS separately (instead of making code generic for all commands, we can have separate handling for HS). 

I can see that I need this flexible/generic code very soon as I am implementing many more commands for rebuild that is why I asked if we can make change now only so that subsequent check-ins do not need to make lots of changes around this area.